### PR TITLE
Jax c51 contrib

### PR DESCRIPTION
**Pull request by kinalmehta**
## Description

JAX implementation for C51
Implementation for #221 

## Types of changes

- [ ] Bug fix
- [ ] New feature
- [x] New algorithm
- [x] Documentation

## Checklist:

- [x] I've read the [CONTRIBUTION](https://github.com/vwxyzjn/cleanrl/blob/master/CONTRIBUTING.md) guide (**required**).
- [x] I have ensured `pre-commit run --all-files` passes (**required**).
- [x] I have updated the documentation and previewed the changes via `mkdocs serve`.
- [ ] I have updated the tests accordingly (if applicable).

If you are adding new algorithms or your change could result in performance difference, you may need to (re-)run tracked experiments. See https://github.com/vwxyzjn/cleanrl/pull/137 as an example PR. 
- [x] I have contacted @vwxyzjn to obtain access to the [openrlbenchmark W&B team](https://wandb.ai/openrlbenchmark) (**required**).
- [x] I have tracked applicable experiments in [openrlbenchmark/cleanrl](https://wandb.ai/openrlbenchmark/cleanrl) with `--capture-video` flag toggled on (**required**).
- [x] I have added additional documentation and previewed the changes via `mkdocs serve`.
    - [x] I have explained note-worthy implementation details.
    - [x] I have explained the logged metrics.
    - [ ] I have added links to the original paper and related papers (if applicable).
    - [ ] I have added links to the PR related to the algorithm.
    - [x] I have created a table comparing my results against those from reputable sources (i.e., the original paper or other reference implementation).
    - [x] I have added the learning curves (in PNG format with `width=500` and `height=300`).
    - [x] I have added links to the tracked experiments.
- [ ] I have updated the tests accordingly (if applicable).

### CODE REVIEW

**Comment by kinalmehta** [comment]
Results on classical gym environments can be checked here.
https://wandb.ai/kinalmehta/jax-cleanrl/reports/C51-JAX-vs-Pytorch-on-Classical-Gym-Environments--VmlldzoyNDQ3OTk5

We see a speed-up of about ~30% in the JAX version compared to Pytorch.

**Comment by kinalmehta** [comment]
Here is the benchmark report on atari environments
https://wandb.ai/kinalmehta/jax-cleanrl/reports/C51-JAX-vs-Pytorch-on-Atari-Environments--VmlldzoyNjkyNDY0

**Important observations:**
- BeamRider performance is bad compared PyTorch version
- Breakout performance almost matches PyTorch variant but is still a bit low comparatively
- For Pong, the performance matches perfectly for 2 seeds, but the reward remains zero for one of the seeds.

Need to look into more detail about the differences between PyTorch and JAX implementations to fix the above mentioned issues.

**Comment by joaogui1** [comment]
How does it compare to Dopamine's version?

**Comment by kinalmehta** [comment]
> How does it compare to Dopamine's version?
I haven't checked Dopamine yet. I will have a look and update here, though it might take some time.

**Comment by vwxyzjn** [comment]
FYI dopamine has a [benchmark](https://google.github.io/dopamine/baselines/atari/plots.html), but its x-axis is not the environment steps... Any clue on how we can compare those results? @joaogui1 
image

**Comment by kinalmehta** [comment]
### Atari Fixed
After months of procrastination and debugging various aspects, I finally stumbled upon the cause of performance degradation.
The incorrect epsilon value caused this performance degradation. I missed this detail and used the default value $10^{-8}$ from optax.
However, the C51-PyTorch version uses ${0.01}/{batch\\_size}$. Hoowever I couldn't find any motivation for using this value.

Reading up more on this led to the conclusion that this is a common issue even in NLP and CV as well.
More about this hyperparameter can be read [here](https://stackoverflow.com/questions/43221065/how-does-the-epsilon-hyperparameter-affect-tf-train-adamoptimizer).

### Benchmarking classical envs on CPU
I have updated the plots of classical gym environments (CartPole, Acrobot, MountainCar) by benchmarking on CPU. We see significant speed-up compared to pytorch version on CPU.

### Comparison with dopamine
Based on the beamrider plot shared above, the below table summarizes the final score comparison
| **implementation** | **score** |
| - | - |
| dopamine | 5000-7000 |
| cleanrl-pytorch | ~9500 |
| cleanrl-jax-old | ~2500 |
| cleanrl-jax-fixed | ~9500 |

### Reports link
- https://wandb.ai/kinalmehta/cleanrl/reports/Regression-Report-c51_atari_jax--VmlldzozMjM1Mzc1
- https://wandb.ai/kinalmehta/cleanrl/reports/Regression-Report-c51_jax--VmlldzozMjM1MzM0

### Conclusion
The updated plots are available on the above links itself.
The PR looks good to be mearged once the documentation is updated. Anything else I am missing here @vwxyzjn?

**Comment by vwxyzjn** [comment]
The results look incredible. Great job @kinalmehta. Thanks for chasing down the cause for the issue. The code also look great to me. Feel free to start adding documentation. You should also move the experiments to the `openrlbenchmark/cleanrl` namespace.

**Comment by kinalmehta** [comment]
I've added the documentation, and now I believe this PR is ready for the final review.

File: docs/rl-algorithms/c51.md [hunk 1]
```diff
@@ -16,7 +16,8 @@ Original papers:
 | ----------- | ----------- |
 | :material-github: [`c51_atari.py`](https://github.com/vwxyzjn/cleanrl/blob/master/cleanrl/c51_atari.py), :material-file-document: [docs](/rl-algorithms/c51/#c51_ataripy) |  For playing Atari games. It uses convolutional layers and common atari-based pre-processing techniques. |
 | :material-github: [`c51.py`](https://github.com/vwxyzjn/cleanrl/blob/master/cleanrl/c51.py), :material-file-document: [docs](/rl-algorithms/c51/#c51py) | For classic control tasks like `CartPole-v1`. |
-
+| :material-github: [`c51_atari.py`](https://github.com/vwxyzjn/cleanrl/blob/master/cleanrl/c51_atari_jax.py), :material-file-document: [docs](/rl-algorithms/c51/#c51_atari_jaxpy) |  For playing Atari games. It uses convolutional layers and common atari-based pre-processing techniques. |
+| :material-github: [`c51.py`](https://github.com/vwxyzjn/cleanrl/blob/master/cleanrl/c51_jax.py), :material-file-document: [docs](/rl-algorithms/c51/#c51_jaxpy) | For classic control tasks like `CartPole-v1`. |
```
Comment (vwxyzjn) [inline docs/rl-algorithms/c51.md:20]: These need to be renamed to c51_atari_jax.py and c51_jax.py.

The overview.md needs to be updated as well.

File: docs/rl-algorithms/c51.md [hunk 2]
```diff
@@ -180,6 +181,129 @@ Tracked experiments and game play videos:
...
+<div class="grid-container">
+<img src="../c51/jax/BeamRiderNoFrameskip-v4.png">
+<img src="../c51/jax/BeamRiderNoFrameskip-v4-time.png">
+
+<img src="../c51/jax/BreakoutNoFrameskip-v4.png">
+<img src="../c51/jax/BreakoutNoFrameskip-v4-time.png">
+
+<img src="../c51/jax/PongNoFrameskip-v4.png">
+<img src="../c51/jax/PongNoFrameskip-v4-time.png">
+</div>
```
Comment (vwxyzjn) [inline docs/rl-algorithms/c51.md:238]: These look great! A minor note for the future: we shouldn't feel obligated to export the wandb curves manually in the future anymore -- the rlops utility should give us the `compare.png` and `compare-time.png` that we can use directly to save manual labor :)

**Review by vwxyzjn — APPROVED** [review]
Some minor comments. Great work! Feel free to merge once they are addressed.

One quick note is that https://github.com/vwxyzjn/cleanrl/pull/292 is getting merged. Would you mind submitting a PR to https://github.com/vwxyzjn/cleanrl/pull/292, adding model eval and uploading trained models to huggingface (for just one random seed)?

File: docs/rl-algorithms/c51.md (hunk 1 above)
Comment (kinalmehta) [reply docs/rl-algorithms/c51.md]: done
